instrumentor/src/index.js

```diff
@@ -57,39 +62,127 @@ const instrumentTouchables = path => {
+  return t.arrowFunctionExpression([t.identifier('e')], functionBody);
+};
+
+const extendsReactComponent = path => {
+  // By the time we start traversing the AST for Switch.js, the 'Switch' class looks like:
+  //
+  //   var Switch = function (_React$Component) {
+  //     (0, _inherits2.default)(Switch, _React$Component);
+  //
+  //     function Switch() {
+  //       <class body>
+  //     }
+  //
+  //     return Switch;
+  //   }(React.Component);
+  //
+  // so to determine whether this extends 'React.Component', we should check the argument to the
+  // function on the RHS of the variable declarator node for 'React.Component' or 'Component'.
+  // :TODO: (jmtaber129): Consider adding a check for the line that looks like:
+  //   (0, _inherits2.default)(Switch, _React$Component)
+  const reactComponentCandidate =
+    path.node.init && path.node.init.arguments && path.node.init.arguments[0];
+  return (
+    reactComponentCandidate &&
+    ((reactComponentCandidate.object &&
+      reactComponentCandidate.object.name === 'React' &&
+      reactComponentCandidate.property &&
+      reactComponentCandidate.property.name === 'Component') ||
+      reactComponentCandidate.name === 'Component')
+  );
+};
+
+const instrumentSwitch = path => {
+  if (instrumentationMap.has('switch')) {
+    // We already instrumented the switch, so do nothing.
+    return;
+  }
+
+  // The method we want to instrument:
+  // * Is named '_handleChange'
+  // * Has a variable declarator parent named 'Switch'
+  // * Extends 'React.Component'.
+  if (
+    path.node.left.property &&
+    path.node.left.property.name === '_handleChange'
+  ) {
+    const parent = path.findParent(path => {
+      return (
+        path.isVariableDeclarator() &&
+        path.node.id.name === 'Switch' &&
+        extendsReactComponent(path)
+      );
     });
 
-    // Call the function body with the event parameter 'e'.
-    const replacementFunc = t.arrowFunctionExpression(
-      [t.identifier('e')],
-      functionBody
+    if (!parent) {
+      return;
+    }
+
+    // Create the expression for calling the original function for this listener.
+    // '(<original function>).call(this, e)'.
+    const originalFunctionExpression = path.node.right;
+    const replacementFunc = getOriginalFunctionReplacement(
+      originalFunctionExpression,
+      '_this',
+      'autotrackSwitchChange',
+      path.node.left.property.name
     );
-    path.get('value').replaceWith(replacementFunc);
+
+    path.get('right').replaceWith(replacementFunc);
+
+    // :KLUDGE: There's some unexpected behavior in the babel traverser that seems to cause the AST
+    // node we're instrumenting to be visited multiple times. To avoid unintentionally wrapping the
+    // same method multiple times, record that we've instrumented the switch.
+    // :TODO: (jmtabe129): Remove this once we figure out what's going on here.
+    instrumentationMap.add('switch');
```

Comment:
Gotcha... not a horrible kludge, but the string should definitely be pulled out into a CONST so we can use it in both spots.

Reply:
Woops, good catch